crates/pallets/starknet/src/lib.rs

```diff
@@ -1039,6 +996,53 @@ impl<T: Config> Pallet<T> {
         frame_system::Pallet::<T>::deposit_log(digest);
     }
 
+    /// Aggregate L2 > L1 messages from the call info.
+    ///
+    /// # Arguments
+    ///
+    /// * `call_info` — A ref to the call info structure.
+    /// * `next_order` — Next expected message order, has to be 0 for a top level invocation
+    ///
+    /// # Returns
+    ///
+    /// Next expected message order
+    fn aggregate_messages_in_call_info(tx_hash: TransactionHash, call_info: &CallInfo, next_order: usize) -> usize {
+        let mut message_idx = 0;
+        let mut inner_call_idx = 0;
+        let mut next_order = next_order;
+
+        loop {
+            // Store current call's messages as long as they have sequential orders
+            if message_idx < call_info.execution.l2_to_l1_messages.len() {
+                let ordered_message = &call_info.execution.l2_to_l1_messages[message_idx];
+                if ordered_message.order == next_order {
+                    let message = MessageToL1 {
+                        from_address: call_info.call.storage_address,
+                        to_address: ordered_message.message.to_address,
+                        payload: ordered_message.message.payload.clone(),
+                    };
+                    TxMessages::<T>::append(tx_hash, message);
+                    next_order += 1;
+                    message_idx += 1;
+                    continue;
+                }
+            }
+
+            // Go deeper to find the continuation of the sequence
+            if inner_call_idx < call_info.inner_calls.len() {
+                next_order =
+                    Self::aggregate_messages_in_call_info(tx_hash, &call_info.inner_calls[inner_call_idx], next_order);
+                inner_call_idx += 1;
```

Comment:
It will force us to do multiple iterations over the array. First collecting the events in correct order, then iterating over and emiting and storing.
With our impl we a only do the iteration one, emiting and storing at the same time.
Does it make sense to keep it that way?

Reply:
Hmm, I guess I get what you're saying. The overall complexity would still be the same though I believe O(n) where n is the total number of call_info. But yes, the current approach will have less iterations. Its easier to understand code vs difficult to understand but slightly efficient. I don't know what the pragmatic thing to do here is so whatever you feel is right :)